engine/src/main/java/org/terasology/config/flexible/FlexibleConfig.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2016 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.config.flexible;
+
+import com.google.common.collect.Maps;
+import org.terasology.assets.ResourceUrn;
+
+import java.util.Map;
+
+public class FlexibleConfig {
+    private Map<Key<?>, Setting> settingMap;
+
+    public FlexibleConfig() {
+        this.settingMap = Maps.newHashMap();
+    }
+
+    public <V> Key<V> add(Setting<V> setting) {
+        Key<V> key = new Key<>(setting.getId(), setting.getValue());
+
+        // Maybe throw an exception?
+        if (has(key))
+            return null;
+
+        settingMap.put(key, setting);
+        return key;
+    }
+
+    public boolean remove(Key<?> id) {
+        Setting setting = get(id);
+
+        if (setting == null || setting.hasSubscribers())
+            return false;
+
+        settingMap.remove(id);
+        return true;
+    }
+
+    @SuppressWarnings("unchecked")
+    public <V> Setting<V> get(Key<V> key) {
+        if (key == null || key.vClass == null) return null;
+
+        Setting setting = settingMap.get(key);
+
+        if (setting == null) return null;
+
+        // Possibly raise an exception?
```

Comment:
Alright. One thing though: Making the user cast it will cause loads of unchecked cast warnings. Instead, I propose to make the `get` method perform the cast for them and suppress the unchecked cast warning for it. Instead of `Setting<Integer> setting = (Setting<Integer>) config.get(uri)` polluting the codebase with unchecked cast warnings, we can have either `Setting<Integer> setting = config.get(uri, Integer.class)` or `Setting<Integer> setting = config.get<Integer>(uri)`. Since we don't have a need for the Class (unless you want to return null instead of raising an exception if the types are invalid), I propose the second method.